File: Data Structures & Algorithms/delete-node-in-a-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(!root)
        {
            return nullptr;
        }
        if(root->val == key)
        {
            // Case 1: no children
            if(!root->left && !root->right) 
            {
                return nullptr;
            }
            // Case 2: one child
            if(!root->left)
            {
                return root->right;
            }
            if(!root->right)
            {
                return root->left;
            }
            // Case 3: two children
            // Step 1: Find successor (go right once, then left all the way)
            TreeNode* successor = root->right;
            while(successor->left)
            {
                successor=successor->left;
            }
            // Step 2: Copy value
            root->val = successor->val;
            root->right = deleteNode(root->right,successor->val);
            return root;
            
        }
        if(key>root->val)
        {
            root->right=deleteNode(root->right,key);
            return root;
        }
        else {
            root->left=deleteNode(root->left,key);
            return root;
        }
    }
};

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, node1: Optional[TreeNode], node2: Optional[TreeNode])->bool:
        if node1==None and node2==None:
            return True
        elif not node1 or not node2 or node1.val!=node2.val:
            return False
        return self.isSameTree(node1.left,node2.left)and self.isSameTree(node1.right,node2.right)

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if subRoot==None:
            return True
        elif root==None and subRoot!=None:
            return False
        elif self.isSameTree(root,subRoot):
            return True
        return self.isSubtree(root.left,subRoot) or self.isSubtree(root.right,subRoot)
        
